src/rrule.ts

```diff
@@ -234,11 +238,51 @@ export default class RRule implements QueryMethods {
     return result as Date
   }
 
+  /**
+   * Returns the very first occurrence of this rrule if it exists, null otherwise
+   *
+   * @returns {Date | null}
+   */
+  first (): Date | null {
+    return this.after(this.options.dtstart, true) || null
+  }
+
+  /**
+   * Returns the last occurrence of this rrule if it exists, null otherwise
+   * @returns {Date | null}
+   */
+  last (): Date | null {
+    if (!this.isFinite()) {
+      return null
+    }
+
+    if (this.options.until) {
+      return this.before(this.options.until, true)
+    }
+
+    const allDates = this.all()
```

Comment:
Suggestion to use this.before() instead of this.all() here with a max date (new Date(9999, 1, 1), or some other date that is large enough to be agreed upon) so that the values from .all() are not all stored into memory while computing the result.

Reply:
Date has a max value which seems to be `new Date(8640000000000000);`, anything after that returns "Invalid Date". We could use that if it's not more expensive (which I think it isn't)

Reply:
Yea, that or a similar value works.

Technically the Iter loop won't works/exits properly with that as the MaxDate in IterResult, since technically tooLate will never be set to true, as `new Date(8640000000000001) > new Date(8640000000000000)` is false.

Would just round the numbers and go with `new Date(8000000000000000);` to avoid that issue. Kind of doubt people are using dates that are in the year 255000+.

That said, the Javascript VM is probably going to long before it gets there if for whatever reason the iteration is that large. Would leave it up to code outside of the library to catch malicious/bad input that would let a massive amount of iterations to occur.

Reply:
Ah yes you're right. But I think it could be a good occasion to fix the exit condition to take NaN dates into account! :)